// @plasmicapp/watcher/dist/index.esm.js
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/watcher.ts
import socketio from "socket.io-client";
var PlasmicRemoteChangeWatcher = class {
  constructor(opts) {
    this.opts = opts;
    this.watchers = [];
    this.socket = void 0;
    var _a;
    this.host = (_a = opts.host) != null ? _a : "https://studio.plasmic.app";
  }
  subscribe(watcher) {
    this.watchers.push(watcher);
    this.ensureWatch();
    return () => {
      this.unsubscribe(watcher);
    };
  }
  unsubscribe(watcher) {
    const index = this.watchers.indexOf(watcher);
    if (index >= 0) {
      this.watchers.splice(index, 1);
      if (this.watchers.length === 0 && this.socket) {
        this.socket.disconnect();
        this.socket = void 0;
      }
    }
  }
  connectSocket() {
    return __async(this, null, function* () {
      const socket = socketio(this.host, {
        path: `/api/v1/socket`,
        transportOptions: {
          polling: {
            extraHeaders: this.makeAuthHeaders()
          }
        }
      });
      return socket;
    });
  }
  makeAuthHeaders() {
    const tokens = this.opts.projects.map((p) => `${p.id}:${p.token}`).join(",");
    return {
      "x-plasmic-api-project-tokens": tokens
    };
  }
  ensureWatch() {
    return __async(this, null, function* () {
      if (this.socket) {
        return;
      }
      if (this.watchers.length === 0) {
        return;
      }
      const socket = this.socket = yield this.connectSocket();
      socket.on("initServerInfo", () => {
        socket.emit("subscribe", {
          namespace: "projects",
          projectIds: this.opts.projects.map((p) => p.id)
        });
      });
      socket.on("error", (data) => {
        console.error(`${(/* @__PURE__ */ new Date()).toISOString()}: Encountered error ${data}`);
        this.watchers.forEach((watcher) => {
          var _a;
          return (_a = watcher.onError) == null ? void 0 : _a.call(watcher, data);
        });
        socket.disconnect();
        this.socket = void 0;
      });
      socket.on("update", (data) => __async(this, null, function* () {
        console.log(
          `${(/* @__PURE__ */ new Date()).toISOString()}: Project ${data.projectId} updated to revision ${data.revisionNum}`
        );
        this.watchers.forEach(
          (watcher) => {
            var _a;
            return (_a = watcher.onUpdate) == null ? void 0 : _a.call(watcher, data.projectId, data.revisionNum);
          }
        );
      }));
      socket.on("publish", (data) => __async(this, null, function* () {
        console.log(
          `${(/* @__PURE__ */ new Date()).toISOString()}: Project ${data.projectId} updated to version ${data.version}`
        );
        this.watchers.forEach(
          (watcher) => {
            var _a;
            return (_a = watcher.onPublish) == null ? void 0 : _a.call(watcher, data.projectId, data.version);
          }
        );
      }));
    });
  }
};
export {
  PlasmicRemoteChangeWatcher
};
